src/tests/redecomp_sparsematrix.cpp

```diff
@@ -0,0 +1,223 @@
+      n_row_entries[i] = 0;
+      for (int j=0; j<n_local_elem; j++) {
+        redecomp_mesh.GetElementCenter(j, xj);
+        // evaluate kernel which takes in both element centroids
+        double Wij = filter_kernel(xi, xj);
+        // if kernel returns a non-zero value, create entry in the sparse redecomp matrix
+        if (Wij > 0.0) {
+          W_redecomp.Add(i, j, Wij);
+          n_row_entries[i]++;
+        }
+      }
+    }
+
+    W_redecomp.Finalize();
+
+    // normalize each row of filter matrix to conserve mass
+    for (int i=0; i<n_local_elem; i++) {
+      mfem::Vector row_data(W_redecomp.GetRowEntries(i), n_row_entries[i]);
+      row_data /= row_data.Norml1();
+    }
+
+    // transfer matrix to mfem::ParMesh
+    auto W = matrix_xfer.TransferToParallel(W_redecomp);
+
+    // define analytical field for testing
+    auto x_function = [&](const mfem::Vector&x) {
+      double pi = 3.1415;
+      double f = 0.0;
+      for (int i=0; i<dim; i++) {
+        f += std::sin(2.*pi*x[i]/side_length);
+      }
+      return 0.5 + 0.5*f/dim;
+    };
+
+    // leverage uniform mesh to write analytical definition of the filtered field
+    auto xf_function = [&](const mfem::Vector&x) {
+      mfem::Vector xj(dim);
+      double value = 0.0;
+      double denom = 0.0;
+      if (dim == 2) {
+        for (int i=0; i<n_elem_per_dim; i++) {
+          xj(0) = dx/2. + i*dx;
+          for (int j=0; j<n_elem_per_dim; j++) {
+            xj(1) = dx/2. + j*dx;
+            value += filter_kernel(x, xj)*x_function(xj);
+            denom += filter_kernel(x, xj);
+          }
+        }
+      }
+      else {
+        for (int i=0; i<n_elem_per_dim; i++) {
+          xj(0) = dx/2. + i*dx;
+          for (int j=0; j<n_elem_per_dim; j++) {
+            xj(1) = dx/2. + j*dx;
+              for (int k=0; k<n_elem_per_dim; k++) {
+                xj(2) = dx/2. + k*dx;
+                value += filter_kernel(x, xj)*x_function(xj);
+                denom += filter_kernel(x, xj);
+            }
+          }
+        }
+      }
+      return value / denom;
+    };
+
+    // compare filtered filed with analytical solution
+    mfem::FunctionCoefficient xCoef(x_function);
+    mfem::FunctionCoefficient xfCoef(xf_function);
+    mfem::ParGridFunction x(&par_fes);
+    mfem::ParGridFunction xf_filt(&par_fes);
+    mfem::ParGridFunction xf_func(&par_fes);
+
+    // evaulate analytical field
+    x.ProjectCoefficient(xCoef);
+
+    // filter field
+    W->Mult(x, xf_filt);
+
+    // compute analytical filtered field
+    xf_func.ProjectCoefficient(xfCoef);
+
+    // compute error
+    mfem::ParGridFunction error  = xf_filt;
+                          error -= xf_func;
+    max_error_ = mfem::InnerProduct(par_mesh.GetComm(), error, error);
+  }
+};
+
+TEST_P(SparseMatrixTest, mass_matrix_transfer)
+{
+  EXPECT_LT(max_error_, 1.0e-13);
+
+  MPI_Barrier(MPI_COMM_WORLD);
+}
+
+INSTANTIATE_TEST_SUITE_P(redecomp, SparseMatrixTest, testing::Values(
```

Comment:
Let's double check test runtime in CI. Let's double check that each refinement level is in fact testing something different. Perhaps refinement 1 and 2 are not testing anything new, but 3 guarantees zero element redecomp meshes, etc.